From 0000000000000000000000000000000000000003 Mon Sep 17 00:00:00 2001
From: Keystone P550 Integration <keystone-p550@example.invalid>
Date: Thu, 20 Aug 2026 00:00:00 +0800
Subject: [PATCH] sm: adapt VF2 SBI dispatch to OpenSBI 1.6

OpenSBI 1.6 returns from traps after the extension handler and supplies a
skip_regs_update flag for handlers which replace the complete register state.
It also replaced the out-value/out-trap callback arguments with
struct sbi_ecall_return. Preserve Keystone's positive SM error-code ABI by
writing a0/a1 explicitly and asking the generic dispatcher not to overwrite
them.

Upstream-Status: Inappropriate [VF2 compatibility for P550]
---
 src/sm-sbi-opensbi.c | 61 ++++++++++++++++++++++++++++++++------------
 src/sm-sbi.c         |  5 ----
 2 files changed, 37 insertions(+), 29 deletions(-)

diff --git a/src/sm-sbi.c b/src/sm-sbi.c
index 4616433..0c61753 100644
--- a/src/sm-sbi.c
+++ b/src/sm-sbi.c
@@ -35,40 +35,35 @@ unsigned long sbi_sm_destroy_enclave(unsigned long eid)
 
 unsigned long sbi_sm_run_enclave(struct sbi_trap_regs *regs, unsigned long eid)
 {
   regs->a0 = run_enclave(regs, (unsigned int) eid);
   regs->mepc += 4;
-  sbi_trap_exit(regs);
   return 0;
 }
 
 unsigned long sbi_sm_resume_enclave(struct sbi_trap_regs *regs, unsigned long eid)
 {
   unsigned long ret;
   ret = resume_enclave(regs, (unsigned int) eid);
   if (!regs->zero)
     regs->a0 = ret;
   regs->mepc += 4;
-
-  sbi_trap_exit(regs);
   return 0;
 }
 
 unsigned long sbi_sm_exit_enclave(struct sbi_trap_regs *regs, unsigned long retval)
 {
   regs->a0 = exit_enclave(regs, cpu_get_enclave_id());
   regs->a1 = retval;
   regs->mepc += 4;
-  sbi_trap_exit(regs);
   return 0;
 }
 
 unsigned long sbi_sm_stop_enclave(struct sbi_trap_regs *regs, unsigned long request)
 {
   regs->a0 = stop_enclave(regs, request, cpu_get_enclave_id());
   regs->mepc += 4;
-  sbi_trap_exit(regs);
   return 0;
 }
 
 unsigned long sbi_sm_attest_enclave(uintptr_t report, uintptr_t data, uintptr_t size)
 {
diff --git a/src/sm-sbi-opensbi.c b/src/sm-sbi-opensbi.c
index e579fda..82d83e6 100644
--- a/src/sm-sbi-opensbi.c
+++ b/src/sm-sbi-opensbi.c
@@ -12,74 +12,87 @@
 #include "pmp.h"
 #include "sm-sbi.h"
 #include "sm.h"
 #include "cpu.h"
 
-static int sbi_ecall_keystone_enclave_handler(unsigned long extid, unsigned long funcid,
-                     const struct sbi_trap_regs *regs,
-                     unsigned long *out_val,
-                     struct sbi_trap_info *out_trap)
+static int sbi_ecall_keystone_enclave_handler(unsigned long extid,
+                     unsigned long funcid,
+                     struct sbi_trap_regs *regs,
+                     struct sbi_ecall_return *out)
 {
   uintptr_t retval;
+  unsigned long out_value = 0;
 
-  if (funcid <= FID_RANGE_DEPRECATED) { return SBI_ERR_SM_DEPRECATED; }
+  (void)extid;
+  out->skip_regs_update = true;
+
+  if (funcid <= FID_RANGE_DEPRECATED) {
+    retval = SBI_ERR_SM_DEPRECATED;
+    goto done;
+  }
   else if (funcid <= FID_RANGE_HOST)
   {
-    if (cpu_is_enclave_context())
-      return SBI_ERR_SM_ENCLAVE_SBI_PROHIBITED;
+    if (cpu_is_enclave_context()) {
+      retval = SBI_ERR_SM_ENCLAVE_SBI_PROHIBITED;
+      goto done;
+    }
   }
   else if (funcid <= FID_RANGE_ENCLAVE)
   {
-    if (!cpu_is_enclave_context())
-      return SBI_ERR_SM_ENCLAVE_SBI_PROHIBITED;
+    if (!cpu_is_enclave_context()) {
+      retval = SBI_ERR_SM_ENCLAVE_SBI_PROHIBITED;
+      goto done;
+    }
   }
 
   switch (funcid) {
     case SBI_SM_CREATE_ENCLAVE:
-      retval = sbi_sm_create_enclave(out_val, regs->a0);
+      retval = sbi_sm_create_enclave(&out_value, regs->a0);
       break;
     case SBI_SM_DESTROY_ENCLAVE:
       retval = sbi_sm_destroy_enclave(regs->a0);
       break;
     case SBI_SM_RUN_ENCLAVE:
-      retval = sbi_sm_run_enclave((struct sbi_trap_regs*) regs, regs->a0);
-      __builtin_unreachable();
-      break;
+      sbi_sm_run_enclave(regs, regs->a0);
+      return 0;
     case SBI_SM_RESUME_ENCLAVE:
-      retval = sbi_sm_resume_enclave((struct sbi_trap_regs*) regs, regs->a0);
-      __builtin_unreachable();
-      break;
+      sbi_sm_resume_enclave(regs, regs->a0);
+      return 0;
     case SBI_SM_RANDOM:
-      *out_val = sbi_sm_random();
+      out_value = sbi_sm_random();
       retval = 0;
       break;
     case SBI_SM_ATTEST_ENCLAVE:
       retval = sbi_sm_attest_enclave(regs->a0, regs->a1, regs->a2);
       break;
     case SBI_SM_GET_SEALING_KEY:
       retval = sbi_sm_get_sealing_key(regs->a0, regs->a1, regs->a2);
       break;
     case SBI_SM_STOP_ENCLAVE:
-      retval = sbi_sm_stop_enclave((struct sbi_trap_regs*) regs, regs->a0);
-      __builtin_unreachable();
-      break;
+      sbi_sm_stop_enclave(regs, regs->a0);
+      return 0;
     case SBI_SM_EXIT_ENCLAVE:
-      retval = sbi_sm_exit_enclave((struct sbi_trap_regs*) regs, regs->a0);
-      __builtin_unreachable();
-      break;
+      sbi_sm_exit_enclave(regs, regs->a0);
+      return 0;
     case SBI_SM_CALL_PLUGIN:
       retval = sbi_sm_call_plugin(regs->a0, regs->a1, regs->a2, regs->a3);
       break;
     default:
       retval = SBI_ERR_SM_NOT_IMPLEMENTED;
       break;
   }
 
-  return retval;
+done:
+  regs->mepc += 4;
+  regs->a0 = retval;
+  regs->a1 = out_value;
+  return 0;
 
 }
 
 struct sbi_ecall_extension ecall_keystone_enclave = {
+  .name = "keystn",
   .extid_start = SBI_EXT_EXPERIMENTAL_KEYSTONE_ENCLAVE,
   .extid_end = SBI_EXT_EXPERIMENTAL_KEYSTONE_ENCLAVE,
+  .experimental = true,
   .handle = sbi_ecall_keystone_enclave_handler,
 };
-- 
2.34.1
